File: androidApp/src/main/java/de/tum/informatics/www1/artemis/native_app/android/service/impl/courses/CourseRegistrationServiceImpl.kt
package de.tum.informatics.www1.artemis.native_app.android.service.impl.courses

import de.tum.informatics.www1.artemis.native_app.android.content.Course
import de.tum.informatics.www1.artemis.native_app.android.content.account.Account
import de.tum.informatics.www1.artemis.native_app.android.service.CourseRegistrationService
import de.tum.informatics.www1.artemis.native_app.android.service.impl.KtorProvider
import de.tum.informatics.www1.artemis.native_app.android.util.NetworkResponse
import de.tum.informatics.www1.artemis.native_app.android.util.performNetworkCall
import io.ktor.client.call.*
import io.ktor.client.request.*
import io.ktor.http.*

class CourseRegistrationServiceImpl(
    private val ktorProvider: KtorProvider
) : CourseRegistrationService {

    override suspend fun fetchRegistrableCourses(
        serverUrl: String,
        authToken: String
    ): NetworkResponse<List<Course>> {
        return performNetworkCall {
            ktorProvider.ktorClient
                .get(serverUrl) {
                    url {
                        appendPathSegments("api", "courses", "for-registration")
                    }

                    bearerAuth(authToken)
                    contentType(ContentType.Application.Json)
                }
                .body()
        }
    }

    override suspend fun registerInCourse(
        serverUrl: String,
        authToken: String,
        courseId: Int
    ): NetworkResponse<Unit> {
        return performNetworkCall {
            val user: Account = ktorProvider.ktorClient
                .post(serverUrl) {
                    url {
                        appendPathSegments("api", courseId.toString(), "register")
                    }

                    bearerAuth(authToken)
                }.body()

            //TODO: sync groups
        }
    }
}

File: iosApp/iosApp/service/impl/CourseRegistrationServiceImpl.swift
import Foundation
import AsyncHTTPClient

class CourseRegistrationServiceImpl: CourseRegistrationService {

    private let jsonProvider: JsonProvider

    init(jsonProvider: JsonProvider) {
        self.jsonProvider = jsonProvider
    }

    func fetchRegistrableCourses(serverUrl: String, authToken: String) async -> NetworkResponse<[Course]> {
        await HTTPClient(eventLoopGroupProvider: .createNew).use { httpClient in
            await performNetworkCall(httpClient: httpClient, maxBodyByteSize: 30.MB(), createRequest: {
                var request = HTTPClientRequest(url: serverUrl + "api/courses/for-registration")
                request.method = .GET

                request.headers.add(name: HttpHeaders.Accept, value: ContentTypes.Application.Json)
                request.headers.add(name: HttpHeaders.UserAgent, value: DefaultHttpHeaderValues.ArtemisUserAgent)
                request.headers.add(name: HttpHeaders.Authorization, value: "Bearer " + authToken)

                return request
            }, decode: { body in
                try jsonProvider.decoder.decode([Course].self, from: body)
            })
        }
    }

    func registerInCourse(serverUrl: String, authToken: String, courseId: Int) async -> NetworkResponse<()> {
        await HTTPClient(eventLoopGroupProvider: .createNew).use { httpClient in
            await performNetworkCall(httpClient: httpClient, maxBodyByteSize: 30.MB(), createRequest: {
                var request = HTTPClientRequest(url: serverUrl + "api/" + String(courseId) + "/register")
                request.method = .POST

                request.headers.add(name: HttpHeaders.UserAgent, value: DefaultHttpHeaderValues.ArtemisUserAgent)
                request.headers.add(name: HttpHeaders.Authorization, value: "Bearer " + authToken)

                return request
            }, decode: { body in
                try jsonProvider.decoder.decode(Account.self, from: body)
            })
                    .bind(f: { () })
        }
    }
}